### ISSUE TITLE
Multiple @ in a transformation line are confused as the start of a new metavariable declaration

### ISSUE THREAD

**Issue by JonathonReinhart** [ahf/cocci-syntax#5]
Sorry for the confusing title, I'm trying to use the terminology from [the grammar](https://coccinelle.gitlabpages.inria.fr/website/docs/main_grammar.html).

An example:
```cocci
// a metavariable declaration (or just rule)
@ rule @
expression x, y
position p1, p2
@@

// transformation section
// note there are two @ in this line
foo(x@p1, y@p2)

// another metavariable declaration
// cocci.vim doesn't recognize this as a CocciGroup region
@@
expression foo
@@
```
(Note that when writing this issue, I marked this code fence as `cocci`, and whatever syntax highlighter GitHub uses seemed to highlight this correctly.)

As you can see in this image, `cocci.vim` thinks that the two `@` are part of a metavariable declaration (rule):
![Current appearance](https://user-images.githubusercontent.com/1916566/206836540-c39c18aa-30a4-47f4-9799-3fe62bbe639f.png)

However, if we require that the metavariable declaration (rule) `@` start at the beginning of the line...
```diff
diff --git a/syntax/cocci.vim b/syntax/cocci.vim
index 9bf7a98..b5bc2d9 100644
--- a/syntax/cocci.vim
+++ b/syntax/cocci.vim
@@ -14,7 +14,7 @@ syn keyword CocciKeywords       statement function local list fresh position ide
 syn keyword CocciKeywords       declaration declarer attribute symbol format assignment contained
 syn keyword CocciKeywords       operator global field initializer initialiser iterator name contained
 
-syn region CocciGroup matchgroup=CocciGroupDelim start="@[^@]*@" end="@@" contains=CocciKeywords
+syn region CocciGroup matchgroup=CocciGroupDelim start="^@[^@]*@" end="^@@" contains=CocciKeywords
 
 syn match CocciLineRemoved      "^-.*"
 syn match CocciLineAdded        "^+.*"
```

...then everything looks fine:

![Corrected appearance](https://user-images.githubusercontent.com/1916566/206836657-da43311c-d131-4961-9ef4-2f0d2ab1e7d0.png)

It's not clear to me from looking at [the grammar](https://coccinelle.gitlabpages.inria.fr/website/docs/main_grammar.html) if the `@` *has* to start at the beginning of a line (I've never been good wtih BNF). However, every Coccinelle example I've ever seen is as such.

**Comment by JonathonReinhart** [comment]
BTW as to what GitHub uses, I followed this trail:
- https://stackoverflow.com/a/8886392
- https://github.com/github/linguist/blob/master/lib/linguist/languages.yml
- https://github.com/github/linguist/blob/master/vendor/README.md
- https://github.com/Alhadis/language-etc/blob/master/grammars/smpl.cson

It looks like [their definition]( https://github.com/Alhadis/language-etc/blob/f6daf7247f1c99bf8c3853ccbbe847755931ca04/grammars/smpl.cson#L104-L105) does indeed anchor the metavariables `@` at the beginning of the line:

```cson
	# Metavariables for transformations
	metavariables:
		name:  "meta.diff.header.metavariables.smpl"
		begin: "^(@).*(@)\\s*$"
		end:   "^(@@)"
```

Also worth noting that they *do* allow an `@` in the rule name, where you exclude `@`.

**Comment by ahf** [comment]
Nice catch, I agree we should get this fixed. Would you be up for creating a pull request with your suggested change?

**Comment by Snape3058** [comment]
Fixed in #6 with 8a5ef9844162dbb414541c424a41662aad0ca4d3
in a similar way to the suggestion.
In addition to `@`, `:` is also excluded to avoid matching code blocks.
Meanwhile, for the end pattern, `@@` is not required at the beginning of a line to match the following case.

```cocci
@@ @@
* bla bla
```

@ahf please close this issue, thanks.
I think what I have done has already covered all the requirements in this issue.

**Comment by ahf** [comment]
Sounds good. Thanks for the work here @Snape3058.
